src/cigar.c

```diff
@@ -8,12 +8,46 @@ void edits_to_cigar(const char *edits, char **cigar)
 {
     size_t n = strlen(edits);
     *cigar = malloc(2 * n + 1); // maximal length for a cigar
+
+
     // Compute the CIGAR string
+    memset(*cigar, 0,2*n+1);
+
+    char cur = edits[0];
+    int no = 1;
+
+    for (int i = 1 ; i <= n ; i++) {
+        if (edits[i] == cur) {
+            no++;
+        }
+        else {
+            char *num = no + '0';
+            strncat(*cigar, &num, 1);
+            strncat(*cigar, &cur, 1);
+            cur = edits[i];
+            no = 1;
+        }
+    }
 }
 
 static inline int count_edits(const char *cigar)
 {
-    return 42; // Work out how long the string will be
+    int n = strlen(cigar);
+    int cum = 0;
+    int len = 0;
+    for (int i = 0 ; i < n ; i++ ) {
+        if (cigar[i] == '0') {
+            if (atoi(&cigar[i]) != 0 ) {
+                len += atoi(&cigar[i]);
+                if (atoi(&cigar[i-1]) != 0 ) {
+                    len += atoi(&cigar[i-1])*9;
+                }
+            }
```

Comment:
I am a bit confused about this. So, you check if you are looking at a zero, and if so, you go and work out the number at the beginning of the string `cigar[i:]`. But why zero? You are unlikely to see numbers that start with zero, and zero is just another digit if it is inside a number. There is nothing conceptually different between `111` and `101` when it comes to numbers; one is just ten larger than the other. Checking `atoi(cigar + i) != 0`  makes slightly more sense, since `atoi()` will return zero when `cigar + i` doesn't start with an integer. But you are still looking at every index in `cigar`, so don't you still risk getting `atoi()` values out of the same number multiple times?

I cannot really suggest fixes because I cannot figure out what the thinking behind this code is. I can test it, of course, and I certainly do not get the correct results. Any string you give the function that doesn't have a zero in it, for example `5M`, will return zero since we never enter the `if`-statement in the loop.

Anyway, here are some hints for moving forward. One option is to give up on `atoi` and use `sscanf` instead. You want to get a number followed by a character, so this will match one block:

```c
    int no_ops;
    char op;
    sscanf(cigar, "%d%c", &no_ops, &op);
```

You don't actually need the operation for this function, so you can leave it out if you add `*` to the format string:

```c
    int no_ops;
    sscanf(cigar, "%d%*c", &no_ops);
```

The problem now is figuring out where the next block is. It you match `k` characters with this `sscanf`, you want to move `cigar` forward by `k`, and with `sscanf` we are lucky that we can get how much it matched. You just need the `%n` format option. It gives you how many characters are matched at that point, so if you place it as the last option in the formatting string you get the block size.

```c
    int no_ops, block_size;
    sscanf(cigar, "%d%*c%n", &no_ops, &block_size);
```

Now, a simple implementation could look like this:

```c
static inline int count_edits(const char *cigar)
{
    int no_ops, block_size, no_edits = 0;
    for (; *cigar; cigar += block_size)
    {
        sscanf(cigar, "%d%*c%n", &no_ops, &block_size);
        no_edits += no_ops;
    }
    return no_edits;
}
```

If you worry about malformed CIGARs, you can do a little checking. The `sscanf` will return how many of the variables in the expression it could pass, and you can use that to see if something went wrong.

```c
static inline int count_edits(const char *cigar)
{
    int no_ops, block_size, no_edits = 0;
    for (; *cigar; cigar += block_size)
    {
        int matched = sscanf(cigar, "%d%*c%n", &no_ops, &block_size);
        if (matched != 1)
        {
            // If we haven't matched the two variables, we have an error.
            printf("We had a parser error!\n");
            return -1;
        }
        no_edits += no_ops;
    }
    return no_edits;
}
```

If you prefer working with `atoi` you can still do that. It gives you the number at the beginning of the string it gets as input, so the only trick to it is to skip to the next block after you have extracted the number. Going character by character is tricking, but jumping from block to block is almost trivial:

```c
static inline const char *skip_digit(const char *buf)
{
    for (; *buf && isdigit(*buf); buf++)
        ;
    return buf;
}

static inline const char *skip_non_digit(const char *buf)
{
    for (; *buf && !isdigit(*buf); buf++)
        ;
    return buf;
}

static inline const char *next_block(const char *buf)
{
    return skip_non_digit(skip_digit(buf));
}

static inline int count_edits(const char *cigar)
{
    int no_edits = 0;
    for (; *cigar; cigar = next_block(cigar))
    {
        no_edits += atoi(cigar);
    }
    return no_edits;
}
```